### DISCUSSION TITLE
How to wait for in flight requests to terminate?

### DISCUSSION THREAD

**Discussion by paymog** [fastify/fastify#5140] [Q&A] [answered]
I'm trying to add graceful shutdown to my fastify server. I run my server in kubernetes which will stop routing requests during the shutdown process and wait for a clean shutdown. As of right now, my server doesn't do the clean shutdown which isn't optimal for my users because they get 500 errors during a server restart/update.

Does fastify offer a built in mechanism to wait for in flight requests to terminate?

**Comment by metcoder95** [comment]
👋 
Calling [fastify.close](https://fastify.dev/docs/latest/Reference/Server#close) should do all the trick. It replies with 503 when closing. If you want the request to be routed normally, you need to set [`return503OnClosing: false`](https://fastify.dev/docs/latest/Reference/Server#factory-return-503-on-closing) when creating the fastify app.

**Reply by paymog** [reply]
oh interesting, I tried that but it didn't seem to work.

I added the following code

```typescript
  let isShuttingDown = false;

  const gracefulShutdown = (signal: number) => {
    if (isShuttingDown) {
      return;
    }
    isShuttingDown = true;
    console.info(`${signal} received. Gracefully shutting down.`);

    // Stop accepting new connections
    fastify.close(() => {
      fastify.log.info("clsoing");
    });

    // Wait 30 seconds for existing connections to finish
    setTimeout(() => {
      console.warn("Forcefully shutting down after 30 seconds.");
      process.exit(1); // Exit with a non-zero code to indicate an issue on shutdown
    }, 30000);
  };

  process.on("SIGTERM", gracefulShutdown);
  process.on("SIGINT", gracefulShutdown);
```

I also added the following endpoint

```ts
  fastify.get("/slow", async (request, reply) => {
    // wait 20 seconds
    await new Promise((resolve) => setTimeout(resolve, 20000));
    reply.send({
      status: "ok",
    });
  });
```

Then I did the following:

1. start my server
2. call the `/slow` endpoint
3. send a SIGINT

I would then see the following output

```
 ❮❮❮ curl localhost:3080/api/slow
curl: (52) Empty reply from server
```

which suggests to me that the server didn't correctly wait for the request to terminate. Here are the logs I see from the server

```
[17:28:01 UTC] INFO: Server listening at http://0.0.0.0:3080
[17:28:08 UTC] INFO: incoming request
    req: {
      "method": "GET",
      "url": "/api/slow",
      "hostname": "localhost:3080",
      "remoteAddress": "127.0.0.1",
      "remotePort": 57548
    }
    reqId: "clokbikwc0000u5p2hz9d1n8z"
^CSIGINT received. Gracefully shutting down.
 ELIFECYCLE  Command failed with exit code 130.
```

I guess I have something misconfigured here but it's not clear to me what that might be.

**Reply by metcoder95** [reply, marked as answer]
Hmm, adjusting your example I was able to fully gracefully exit:
```js
let isShuttingDown = false

// Did this so that fetch doesn't force the process to stay alive
// because of keep alive
const fastify = Fastify({ logger: true, keepAliveTimeout: 1000 })

fastify.get('/slow', async (request, reply) => {
  // wait 20 seconds
  await new Promise((resolve) => setTimeout(resolve, 3000))
  reply.send({
    status: 'ok'
  })
})

await fastify.listen()

const gracefulShutdown = async (signal) => {
  if (isShuttingDown) {
    return
  }
  isShuttingDown = true
  console.info(`${signal} received. Gracefully shutting down.`)

  // Wait 30 seconds for existing connections to finish
  const timeout = setTimeout(() => {
    console.warn('Forcefully shutting down after 20 seconds.')
    process.exit(1) // Exit with a non-zero code to indicate an issue on shutdown
  }, 10000)

  timeout.unref()

  // Stop accepting new connections
  await fastify.close()

  console.log('Closed out remaining connections.')
  clearTimeout(timeout)

  process.exit(0)
}

process.on('SIGTERM', gracefulShutdown)
process.on('SIGINT', gracefulShutdown)

fetch(
  `http://localhost:${fastify.server.address().port}/slow`)

setTimeout(() => {
  console.log('Sending sigterm')
  process.kill(process.pid, 'SIGTERM')
}, 1000)
```

Now, it is important to remind you that `keep-alive` connections might be causing your server to take longer than expected to shutdown, this is something I would like to checkout on `fastify`, as we might be missing something that its causing you to `idle` connections do not close when `close` has been called while processing a request, and the connection is not dropped after that; leading to the graceful shutdown timeout.

**Reply by paymog** [reply]
ah, I think the key insight here is that I wasn't `await`ing the fall to `fastify.close`! Thank you for all your help, this has solved my issue 🙂

**Reply by paymog** [reply]
Interestingly, it seem this isn't properly respected in kubernetes. Here are the logs I see when running this code in kubernetes and killing a pod

```
goldsky-api-854d9d57cf-vg4x7 goldsky-api  ELIFECYCLE  Command failed.
goldsky-api-854d9d57cf-vg4x7 goldsky-api {"level":"info","time":1699354473236,"pid":95,"msg":"SIGTERM received. Gracefully shutting down."}
goldsky-api-854d9d57cf-vg4x7 goldsky-api  ELIFECYCLE  Command failed.
```

**Reply by paymog** [reply]
ah, turns out the above is because https://github.com/pnpm/pnpm/issues/2653 - I was starting my docker container in k8s with pnpm instead of calling node directly.

**Comment by tmm1** [comment]
Other packages like `http-graceful-shutdown` [will set the `Connection: close` header](https://github.com/sebhildebrandt/http-graceful-shutdown/blob/dd23f5adf770ffc9d043a29907337aa296e938b7/lib/index.js#L120-L121) on any open requests, to ensure that keep-alive connections do not receive new requests.

This is due to https://github.com/nodejs/node/issues/2642 and covered in detail on https://www.dashlane.com/blog/implementing-nodejs-http-graceful-shutdown

Additionally for http2 requests it is necessary to call `http2session.close()`, which is described in https://github.com/fastify/fastify/issues/5716 and fixed in https://github.com/fastify/fastify/pull/6137
